Goldtree/Goldtree/Program.cs

```diff
@@ -86,204 +83,228 @@ public static void Initialize()
+                USB.Write(c);
+                USB.Write((uint)nspname.Length);
+                USB.Write(nspname);
+                Log.Log(Messages.GoldleafNSPApproval);
+
+                Command rc2 = USB.Read();
+                if (!rc2.MagicOk())
+                {
+                    this.QuitWithError(Messages.USBInvalidCommandReceived);
+                }
+                if (rc2.IsCommandId(CommandId.Finish))
+                {
+                    this.QuitWithError(Messages.GoldleafCancelledInstall);
+                }
+                if (!rc2.IsCommandId(CommandId.Start))
                 {
-                    if(rc.IsCommandId(CommandId.ConnectionResponse))
+                    this.QuitWithError(Messages.USBInvalidCommandReceived);
+                }
+
+                this.SendNSPOverUSB(nsp);    
+            }
+            catch
+            {
+                this.QuitWithError(Messages.NSPGeneralError);
+            }
+        }
+
+        private void SendNSPOverUSB(string nsp)
+        {
+            Log.Log(Messages.GoldleafStartingInstall);
+            try
+            {
+                FileStream fs = new FileStream(nsp, FileMode.Open);
+                StreamStorage ist = new StreamStorage(fs, true);
+                Pfs pnsp = new Pfs(ist);
+                ist.Dispose();
+                fs.Close();
+                fs = new FileStream(nsp, FileMode.Open);
+                uint filecount = (uint)pnsp.Files.Length;
+                Command c = new Command(CommandId.NSPData);
+                USB.Write(c);
+                USB.Write(filecount);
+                int tikidx = -1;
+                int certidx = -1;
+                int tmpidx = 0;
+                foreach (var file in pnsp.Files)
+                {
+                    ulong offset = (ulong)pnsp.HeaderSize + (ulong)file.Offset;
+                    ulong size = (ulong)file.Size;
+                    uint len = (uint)file.Name.Length;
+                    USB.Write(len);
+                    USB.Write(file.Name);
+                    USB.Write(offset);
+                    USB.Write(size);
+                    if (Path.GetExtension(file.Name).Replace(".", "").ToLower() == "tik")
                     {
-                        Log.Log("Connection was established with Goldleaf.");
-                        Log.Log("Select the NSP to send to Goldleaf on the dialog.");
-                        OpenFileDialog fd = new OpenFileDialog()
-                        {
-                            Title = "Select NSP to send to Goldleaf via USB",
-                            Filter = "NSP / Nintendo Submission Package (*.nsp)|*.nsp",
-                            Multiselect = false,
-                        };
-                        if(fd.ShowDialog() == DialogResult.OK)
+                        tikidx = tmpidx;
+                    }
+                    else if (Path.GetExtension(file.Name).Replace(".", "").ToLower() == "cert")
```

Comment:
Probably best to use `string.Compare(a, b, StringComparison.OrdinalIgnoreCase)` over spending cycles performing a `.ToLower()` here.